Python/bytecodes.c

```diff
@@ -4112,23 +4112,31 @@ dummy_func(
             _CALL_BUILTIN_CLASS +
             _CHECK_PERIODIC;
 
-        op(_CALL_BUILTIN_O, (callable, self_or_null, args[oparg] -- res)) {
-            /* Builtin METH_O functions */
+        op(_GUARD_CALL_BUILTIN_O, (callable, self_or_null, unused[oparg] -- callable, self_or_null, unused[oparg])) {
             PyObject *callable_o = PyStackRef_AsPyObjectBorrow(callable);
-
-            int total_args = oparg;
-            if (!PyStackRef_IsNull(self_or_null)) {
-                args--;
-                total_args++;
-            }
+            int total_args = oparg + (!PyStackRef_IsNull(self_or_null));
             EXIT_IF(total_args != 1);
             EXIT_IF(!PyCFunction_CheckExact(callable_o));
             EXIT_IF(PyCFunction_GET_FLAGS(callable_o) != METH_O);
+        }
```

Comment:
You might consider breaking this in two: one guard for the number of args, and one guard for callable type/flags. Could be there are situations where you could remove one but not the other.

Speaking of which, is it equally common to have `NULL` or `self` for these two instructions? Or is it more likely that it's always one or the other?

It might be worth tightening up the specialization for only one of the cases, since that makes the guards much simpler (and easier to remove).